ziro-export-risport.ps1

```diff
@@ -0,0 +1,227 @@
+               <first>${First}</first>
+            </ns:listPhone>
+         </soapenv:Body>
+      </soapenv:Envelope>
+"@
+      
+      return Invoke-RestMethod -Method 'Post' -Authentication Basic -Credential $this.Credential -Headers $headers -Uri $uri -Body $requestBody -SkipCertificateCheck
+   }
+   
+   [System.Xml.XmlDocument] SelectCmDevice([string[]]$DeviceNames) {
+      $uri = "{0}/realtimeservice2/services/RISService70" -f $this.BaseUrl
+      $headers = @{
+         'Content-Type' = 'text/xml; charset=UTF-8'
+      }  
+
+      $bodySelectItemsList = $deviceNames | ForEach-Object { @"
+         <soap:item>
+            <soap:Item>${PSItem}</soap:Item>
+         </soap:item>
+"@}
+
+      $requestBody = @"
+      <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:soap="http://schemas.cisco.com/ast/soap">
+         <soapenv:Header/>
+         <soapenv:Body>
+            <soap:selectCmDevice>
+               <soap:StateInfo></soap:StateInfo>
+               <soap:CmSelectionCriteria>
+                  <soap:MaxReturnedDevices>2000</soap:MaxReturnedDevices>
+                  <soap:DeviceClass>Phone</soap:DeviceClass>
+                  <soap:Model>255</soap:Model>
+                  <soap:Status>Any</soap:Status>
+                  <soap:NodeName></soap:NodeName>
+                  <soap:SelectBy>Name</soap:SelectBy>
+                  <soap:SelectItems>
+                     ${bodySelectItemsList}
+                  </soap:SelectItems>
+                  <soap:Protocol>Any</soap:Protocol>
+                  <soap:DownloadStatus>Any</soap:DownloadStatus>
+               </soap:CmSelectionCriteria>
+            </soap:selectCmDevice>
+         </soapenv:Body>
+      </soapenv:Envelope>
+"@
+      
+      $maxNumAttemptsForRateLimitError = 7
+
+      for ($attempt = 1; $attempt -le $maxNumAttemptsForRateLimitError; $attempt++) {
+         try {
+            return Invoke-RestMethod -Method 'Post' -Authentication Basic -Credential $this.Credential -Headers $headers -Uri $uri -Body $requestBody -SkipCertificateCheck
+         }
+         catch [Microsoft.PowerShell.Commands.HttpResponseException] {
+            # Handle rate limit errors
+            if ($PSItem.ErrorDetails.Message.Contains('Exceeded allowed rate for Rea')) {
+               if ($attempt -lt $maxNumAttemptsForRateLimitError) {
+                  Write-Host "Encountered rate limit error when getting device registration statuses [attempt=${attempt}]. Retrying in 10 sec..."
+                  Start-Sleep -Seconds 10
+                  continue
+               }
+               else {
+                  Write-Host -ForegroundColor Red "Reached max. attempts ${maxNumAttemptsForRateLimitError} for rate limit error when getting device registration statuses"
+                  throw
+               }
+            }
+
+            throw
+         }
+      }
+
+      throw "An unexpected error occured when getting device registration statuses"
```

Comment:
Nitpick: correct `occured` to `occurred`

Reply:
👍